### ISSUE TITLE
parse() should throw instead of returning undefined

### ISSUE THREAD

**Issue by stefanpl** [zeit/ms#123]
What's the reasoning behind `parse()` returning `undefined` when given invalid input (e.g. '1 secnd')? In my opinion, an error will be much easier to detect and feels syntactically more appropriate – passing an invalid string is an erroneous usage of the function, after all.

I'd totally be willing to draft a PR if there's agreement an error should be thrown.

**Comment by drazisil** [comment]
It's unclear to me where this occurs.  https://github.com/vercel/ms/blob/40fdbd5f03bde1ca09aff2884424f92b41061591/src/index.ts#L94 appears to be the only possible location, is this still an issue?

**Comment by meyfa** [comment]
The handling of failure cases still seems sub-optimal to me. This is re: https://github.com/vercel/ms/issues/184#issuecomment-1213069870.

TL;DR: `ms()` throws for inputs that are neither of type `string` nor `number`; additionally, it throws for empty strings, and for non-finite numbers. For strings that aren't empty but that cannot be parsed, `ms()` will return `NaN` (as mentioned above). This complicates error handling unnecessarily.

Given that a major-version release is coming up anyway, I'd recommend to remedy this situation by being consistent: either always throwing, or always returning a specific error value. Additionally, there needs to be documentation in the README on how errors will be indicated to the caller. The TypeScript-specific examples there are unsafe since they don't check for `NaN`.

Side note: `NaN` is not a good choice for an error value IMHO because it is also of type `number`. If the error value were `undefined`, for example, TypeScript (and IDEs) would remind programmers that they should handle this case. This will prevent many non-obvious bugs and security risks especially when `ms()` is used for parsing user input.

Perhaps someone at Vercel could comment what they want to do? I'd be happy to contribute any required code/docs changes.

**Comment by quezak** [comment]
It's especially confusing that it **does** throw when passed an empty string:
```
> ms('')
[redacted]/node_modules/ms/index.js:34
  throw new Error(
  ^

Uncaught Error: val is not a non-empty string or a valid number. val=""
```

And then, despite the error message above, it's **not** thrown when val is not a valid number:
```
> ms('wat')
undefined
```

Could we have a fix to make these cases consistent?

**Comment by dimitropoulos** [comment]
Looping back since 2022 when this was last updated:

## Passing nothing (or `undefined`)

`ms()` is a validation error

<img width="1102" height="194" alt="Image" src="https://github.com/user-attachments/assets/195a2557-fd06-422c-b351-821967b6636c" />

`parse()` is too, but more specific
<img width="1198" height="194" alt="Image" src="https://github.com/user-attachments/assets/2091c8c9-0224-4be5-8e12-ee08f5a4f5a8" />

(for clarity, both of those are the same if you pass undefined)

<img width="1181" height="399" alt="Image" src="https://github.com/user-attachments/assets/0d6f4bc7-2c0c-49ed-8260-258788b934b8" />

## Empty string

The validation error with emptystring is consistent, too:

<img width="1204" height="394" alt="Image" src="https://github.com/user-attachments/assets/ff4a286f-ae98-45c2-b4a3-5feaa996987f" />

## Unrecognized inputs

However, when there's a _parsing_ error, this package returns `NaN` whether you call `parse` directly, or via `ms`.

<img width="258" height="201" alt="Image" src="https://github.com/user-attachments/assets/bc94ef0a-48ef-4940-b948-32259fda2e00" />

---

I don't have full context on why this was done this way, but that's how it's designed and implemented.

Since the original post (of `parse()` returning undefined) is no longer the behavior, I'm closing this issue.  If you'd like to make the case for throwing on unrecognized inputs (perhaps as an option?) then please open a new issue.
